File: packages/react-native-airfob/ios/AirfobSdk.swift
import Foundation

/// The single seam between this package and MOCA System's Airfob SDK.
/// Nothing outside this file imports a MOCA symbol, which is what makes P1
/// shippable without a licence.

public struct AirfobCard {
    public let id: String
    public let name: String
    public let siteName: String
    public let status: String
    public let accessLevels: [String]

    public var asDictionary: [String: Any] {
        [
            "id": id,
            "name": name,
            "siteName": siteName,
            "status": status,
            "accessLevels": accessLevels
        ]
    }
}

public struct AirfobSdkStatus {
    public let sdkReady: Bool
    public let registered: Bool
    public let licence: String
    public let cardCount: Int
}

public struct AirfobError: Error {
    public let code: String
    public let message: String

    public init(_ code: String, _ message: String) {
        self.code = code
        self.message = message
    }
}

public typealias SdkEmitter = (String, [String: Any]) -> Void

public protocol AirfobSdk {
    /// Called at launch. Must be cheap and must not throw when unenrolled.
    func boot(config: [String: Any]) throws
    func status() -> AirfobSdkStatus
    func register(token: String) throws -> [AirfobCard]
    func cards() -> [AirfobCard]
    /// Returns "opened" | "noReader" | "denied"
    func unlock(cardId: String?) throws -> String
    func unregister(cardId: String?) throws -> [AirfobCard]

    /// Clears cached per-reader RSSI values. The SDK learns a signal baseline per
    /// reader; when a reader is moved or replaced that baseline goes stale and
    /// taps start failing at the old distance. MOCA added an API for this in
    /// 2.3.15 — it is the first thing to try when one door goes bad.
    func resetRssi() throws

    func teardown()
}

// MARK: - mock

/// Behaves like the real thing minus the radio. Credentials persist across
/// launches so the enrolment flow can be tested properly.
public final class MockAirfobSdk: AirfobSdk {

    private let emit: SdkEmitter
    private let defaults = UserDefaults.standard
    private let storageKey = "airfob.mock.cardId"

    private var ready = false
    private var stored: [AirfobCard] = []

    public init(emit: @escaping SdkEmitter) {
        self.emit = emit
    }

    public func boot(config: [String: Any]) throws {
        ready = true
        restore()
        AirfobLog.shared.write(
            "info", "sdk", "BOOT",
            "Mock SDK booted with \(stored.count) credential(s)",
            data: ["siteId": config["siteId"] as? String ?? "none"]
        )
        emit("status", ["status": statusDictionary()])
    }

    public func status() -> AirfobSdkStatus {
        AirfobSdkStatus(
            sdkReady: ready,
            registered: !stored.isEmpty,
            licence: "valid",
            cardCount: stored.count
        )
    }

    public func register(token: String) throws -> [AirfobCard] {
        guard ready else { throw AirfobError("E_NOT_READY", "boot() has not been called") }

        stored = [
            AirfobCard(
                id: "mock-" + String(token.prefix(8)),
                name: "Mock Credential",
                siteName: "Mock Site",
                status: "active",
                accessLevels: ["All doors"]
            )
        ]
        persist()
        AirfobLog.shared.write("info", "sdk", "REGISTER", "Registered \(stored.count) credential(s)")
        emit("status", ["status": statusDictionary()])
        return stored
    }

    public func cards() -> [AirfobCard] { stored }

    public func unlock(cardId: String?) throws -> String {
        guard ready else { throw AirfobError("E_NOT_READY", "boot() has not been called") }
        guard !stored.isEmpty else { throw AirfobError("E_NO_CARD", "No credential registered") }

        let rssi = -62
        AirfobLog.shared.write(
            "info", "ble", "UNLOCK", "Door opened",
            data: ["cardId": cardId ?? stored[0].id, "rssi": rssi]
        )
        emit("unlockResult", ["result": "opened", "readerId": "mock-reader", "rssi": rssi])
        return "opened"
    }

    public func unregister(cardId: String?) throws -> [AirfobCard] {
        stored.removeAll()
        persist()
        AirfobLog.shared.write("info", "sdk", "UNREGISTER", "Credentials cleared")
        emit("status", ["status": statusDictionary()])
        return stored
    }

    public func resetRssi() throws {
        AirfobLog.shared.write("info", "ble", "RSSI_RESET", "Cached reader RSSI values cleared")
    }

    public func teardown() { ready = false }

    private func statusDictionary() -> [String: Any] {
        let s = status()
        return [
            "sdkReady": s.sdkReady,
            "registered": s.registered,
            "licence": s.licence,
            "cardCount": s.cardCount
        ]
    }

    private func persist() {
        defaults.set(stored.first?.id, forKey: storageKey)
    }

    private func restore() {
        guard let id = defaults.string(forKey: storageKey) else { return }
        stored = [AirfobCard(id: id, name: "Mock Credential", siteName: "Mock Site",
                             status: "active", accessLevels: ["All doors"])]
    }
}

// MARK: - core

/// Process-wide owner of the SDK instance.
///
/// Exists for one reason: tap-and-go must work while the app is closed. If the
/// SDK were created by the React Native module it would only live as long as the
/// JS layer, which is almost never at the moment a user holds their phone to a
/// reader. So the SDK is started at launch and the RN module attaches to it.
@objc(AirfobCore)
public final class AirfobCore: NSObject {

    @objc public static let shared = AirfobCore()

    private var sdkInstance: AirfobSdk?
    private var emitter: SdkEmitter?
    private var booted = false
    private let lock = NSLock()

    /// Real adapter when the licensed framework is vendored, mock otherwise.
    ///
    /// AIRFOB_SDK is set by the podspec, which switches it on only when an
    /// .xcframework is present in ios/Frameworks — so there is no source edit to
    /// remember, and a build without a licence still compiles.
    private func create() -> AirfobSdk {
        let emitter: SdkEmitter = { [weak self] name, data in self?.dispatch(name, data) }
        #if AIRFOB_SDK
        AirfobLog.shared.write("info", "sdk", "ADAPTER", "Using RealAirfobSdk")
        return RealAirfobSdk(emit: emitter)
        #else
        AirfobLog.shared.write("info", "sdk", "ADAPTER", "No Airfob SDK in this build — using MockAirfobSdk")
        return MockAirfobSdk(emit: emitter)
        #endif
    }

    @objc public var isMock: Bool {
        sdk() is MockAirfobSdk
    }

    public func sdk() -> AirfobSdk {
        lock.lock()
        defer { lock.unlock() }
        if let existing = sdkInstance { return existing }
        let created = create()
        sdkInstance = created
        return created
    }

    /// Called at launch. Guarded so an unenrolled device does nothing expensive:
    /// no scanning, no background mode, no battery cost.
    @objc public func bootIfEnrolled() {
        AirfobLog.shared.attach()
        AirfobLog.shared.write("info", "service", "PROCESS_START", "Airfob core attached")

        guard !sdk().cards().isEmpty else {
            AirfobLog.shared.write(
                "info", "service", "BOOT_SKIP",
                "No credential on this device — not starting the scanner"
            )
            return
        }
        boot(config: [:])
    }

    public func boot(config: [String: Any]) {
        lock.lock()
        let alreadyBooted = booted
        lock.unlock()
        guard !alreadyBooted else { return }

        do {
            try sdk().boot(config: config)
            lock.lock(); booted = true; lock.unlock()
            AirfobLog.shared.write("info", "service", "BOOT", "Scanner started")
        } catch {
            let message = (error as? AirfobError)?.message ?? error.localizedDescription
            AirfobLog.shared.write("error", "service", "BOOT_FAIL", message)
        }
    }

    /// The RN module registers itself here; before that, events go to the log only.
    public func setEmitter(_ next: SdkEmitter?) {
        emitter = next
    }

    private func dispatch(_ name: String, _ data: [String: Any]) {
        if let emitter = emitter {
            emitter(name, data)
        } else {
            AirfobLog.shared.write("debug", "sdk", "EVENT_DROPPED", "No JS listener for \"\(name)\"")
        }
    }
}

File: packages/react-native-airfob/android/src/main/java/com/airfob/AirfobSdk.kt
package com.airfob

import android.content.Context

/**
 * The single seam between this package and MOCA System's Airfob SDK.
 *
 * Nothing outside this file imports a MOCA class. Swapping [MockAirfobSdk] for
 * [RealAirfobSdk] is a one-line change in [AirfobModule], which is what makes
 * P1 shippable without a licence.
 */

data class Card(
    val id: String,
    val name: String,
    val siteName: String,
    val status: String,
    val accessLevels: List<String> = emptyList()
)

data class SdkStatus(
    val sdkReady: Boolean,
    val registered: Boolean,
    val licence: String,
    val cardCount: Int
)

class AirfobError(val code: String, message: String) : Exception(message)

/** Emits an event up to JS. */
typealias SdkEmitter = (name: String, data: Map<String, Any?>) -> Unit

interface AirfobSdk {
    /** Called at process start. Must be cheap and must not throw when unenrolled. */
    fun boot(config: Map<String, Any?>)
    fun status(): SdkStatus
    fun register(token: String): List<Card>
    fun cards(): List<Card>
    /** @return "opened" | "noReader" | "denied" */
    fun unlock(cardId: String?): String
    fun unregister(cardId: String?): List<Card>

    /**
     * Clears cached per-reader RSSI values. The SDK learns a signal baseline for
     * each reader; when a reader is moved or replaced that baseline goes stale
     * and taps start failing at the old distance. MOCA added an API for this in
     * 2.3.15 — it is the first thing to try when one door goes bad.
     */
    fun resetRssi()

    fun teardown()
}

/* ------------------------------------------------------------------ mock --- */

/**
 * Behaves like the real thing minus the radio. Credentials persist across
 * restarts so the enrolment flow can be tested properly.
 */
class MockAirfobSdk(
    private val context: Context,
    private val emit: SdkEmitter
) : AirfobSdk {

    private val prefs by lazy {
        context.getSharedPreferences("airfob-mock", Context.MODE_PRIVATE)
    }

    private var ready = false
    private var stored = mutableListOf<Card>()

    override fun boot(config: Map<String, Any?>) {
        ready = true
        restore()
        AirfobLog.write(
            "info", "sdk", "BOOT",
            "Mock SDK booted with ${stored.size} credential(s)",
            mapOf("siteId" to config["siteId"])
        )
        emit("status", mapOf("status" to statusMap()))
    }

    override fun status() = SdkStatus(
        sdkReady = ready,
        registered = stored.isNotEmpty(),
        licence = "valid",
        cardCount = stored.size
    )

    override fun register(token: String): List<Card> {
        if (!ready) throw AirfobError("E_NOT_READY", "boot() has not been called")
        stored = mutableListOf(
            Card(
                id = "mock-${token.take(8)}",
                name = "Mock Credential",
                siteName = "Mock Site",
                status = "active",
                accessLevels = listOf("All doors")
            )
        )
        persist()
        AirfobLog.write("info", "sdk", "REGISTER", "Registered ${stored.size} credential(s)")
        emit("status", mapOf("status" to statusMap()))
        return stored
    }

    override fun cards(): List<Card> = stored

    override fun unlock(cardId: String?): String {
        if (!ready) throw AirfobError("E_NOT_READY", "boot() has not been called")
        if (stored.isEmpty()) throw AirfobError("E_NO_CARD", "No credential registered")

        val rssi = -62
        AirfobLog.write(
            "info", "ble", "UNLOCK", "Door opened",
            mapOf("cardId" to (cardId ?: stored.first().id), "rssi" to rssi)
        )
        emit("unlockResult", mapOf("result" to "opened", "readerId" to "mock-reader", "rssi" to rssi))
        return "opened"
    }

    override fun unregister(cardId: String?): List<Card> {
        stored.clear()
        persist()
        AirfobLog.write("info", "sdk", "UNREGISTER", "Credentials cleared")
        emit("status", mapOf("status" to statusMap()))
        return stored
    }

    override fun resetRssi() {
        AirfobLog.write("info", "ble", "RSSI_RESET", "Cached reader RSSI values cleared")
    }

    override fun teardown() {
        ready = false
    }

    private fun statusMap(): Map<String, Any?> = status().let {
        mapOf(
            "sdkReady" to it.sdkReady,
            "registered" to it.registered,
            "licence" to it.licence,
            "cardCount" to it.cardCount
        )
    }

    private fun persist() {
        prefs.edit().putString("cardId", stored.firstOrNull()?.id).apply()
    }

    private fun restore() {
        val id = prefs.getString("cardId", null) ?: return
        stored = mutableListOf(Card(id, "Mock Credential", "Mock Site", "active", listOf("All doors")))
    }
}

/* --------------------------------------------------------------- factory --- */

/**
 * Lets [AirfobCore] obtain a real adapter without the main source set naming a
 * class that may not be compiled. RealAirfobSdkFactory lives in src/withSdk and
 * exists only when the licensed AAR is in libs/; [AirfobCore] looks it up by
 * name and falls back to the mock.
 *
 * Implementations must have a no-argument constructor.
 */
interface AirfobSdkFactory {
    fun create(context: Context, emit: SdkEmitter): AirfobSdk
}
